packages/flutter/lib/src/gestures/resampler.dart

```diff
@@ -0,0 +1,424 @@
+        //
+        // When `sampleTime == _next.timeStamp`, all events with
+        // `event.timeStamp > sampleTime` must also have
+        // `event.timeStamp > _next.timeStamp` so no events will be processed
+        // early.
+        //
+        // When the input frequency is no greater than the sampling
+        // frequency, this early processing should guarantee that `up` and
+        // `remove` events are always re-sampled.
+        final Duration nextTimeStamp = _next?.timeStamp ?? const Duration();
+        assert(event.timeStamp >= nextTimeStamp);
+        if (event.timeStamp > nextTimeStamp) {
+          break;
+        }
+      }
+
+      final bool wasTracked = _isTracked;
+      final bool wasDown = _isDown;
+
+      // Update pointer state.
+      _isTracked = event is! PointerRemovedEvent;
+      _isDown = event.down;
+
+      // Position at `sampleTime`.
+      final Offset position = _positionAt(sampleTime);
+
+      // Initialize position if we are starting to track this pointer.
+      if (_isTracked && !wasTracked) {
+        _position = position;
+      }
+
+      // Current pointer identifier.
+      final int pointerIdentifier = event.pointer;
+
+      // Initialize pointer identifier for `move` events.
+      // Identifier is expected to be the same while `down`.
+      assert(!wasDown || _pointerIdentifier == pointerIdentifier);
+      _pointerIdentifier = pointerIdentifier;
+
+      // Skip `move` and `hover` events as they are automatically
+      // generated when the position has changed.
+      if (event is! PointerMoveEvent && event is! PointerHoverEvent) {
+        dequeuedEvents.add(_clone(event, position, position - _position,
+            pointerIdentifier, sampleTime));
+        _position = position;
+      }
+
+      _queuedEvents.removeFirst();
+    }
+
+    return dequeuedEvents;
+  }
+
+  List<PointerEvent> _samplePointerPositions(Duration sampleTime) {
+    final List<PointerEvent> sampledEvents = <PointerEvent>[];
+
+    // Position at `sampleTime`.
+    final Offset position = _positionAt(sampleTime);
+
+    // Add `move` or `hover` events if position changed.
+    final PointerEvent? next = _next;
+    if (position != _position && next != null) {
+      final Offset delta = position - _position;
+      final PointerEvent event = _isDown
+          ? _toMoveEvent(next, position, delta, _pointerIdentifier, sampleTime)
+          : _toHoverEvent(next, position, delta, sampleTime);
+      sampledEvents.add(event);
+      _position = position;
+    }
+
+    return sampledEvents;
+  }
+
+  /// Enqueue pointer `event` for resampling.
+  void addEvent(PointerEvent event) {
+    _queuedEvents.add(event);
+  }
+
+  /// Dequeue resampled pointer events for the specified `sampleTime`.
+  /// This will return multiple events if position is not the only
+  /// state that changed since last sample.
+  List<PointerEvent> sample(Duration sampleTime) {
+    // Process events for `sampleTime`.
```

Comment:
Can we be a little more specific for the comment? `Process events for sampleTime.` sounds already implied by the name of the method.

Reply:
Removed it.